dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,351 @@
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<TRecord> GetBatchAsync(
+        IEnumerable<string> keys,
+        GetRecordOptions? options = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(keys);
+
+        const string OperationName = "Find";
+
+        using var cursor = await this._mongoCollection
+            .FindAsync(this.GetFilterByIds(keys), cancellationToken: cancellationToken)
+            .ConfigureAwait(false);
+
+        while (await cursor.MoveNextAsync(cancellationToken).ConfigureAwait(false))
+        {
+            foreach (var record in cursor.Current)
+            {
+                if (record is not null)
+                {
+                    yield return VectorStoreErrorHandler.RunModelConversion(
+                        DatabaseName,
+                        this.CollectionName,
+                        OperationName,
+                        () => this._mapper.MapFromStorageToDataModel(record, new()));
+                }
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    public Task<string> UpsertAsync(TRecord record, UpsertRecordOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(record);
+
+        const string OperationName = "ReplaceOne";
+
+        var replaceOptions = new ReplaceOptions { IsUpsert = true };
+        var storageModel = VectorStoreErrorHandler.RunModelConversion(
+            DatabaseName,
+            this.CollectionName,
+            OperationName,
+            () => this._mapper.MapFromDataToStorageModel(record));
+
+        var key = storageModel[AzureCosmosDBMongoDBConstants.MongoReservedKeyPropertyName].AsString;
+
+        return this.RunOperationAsync(OperationName, async () =>
+        {
+            var result = await this._mongoCollection
+                .ReplaceOneAsync(this.GetFilterById(key), storageModel, replaceOptions, cancellationToken)
+                .ConfigureAwait(false);
+
+            return key;
+        });
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> UpsertBatchAsync(
+        IEnumerable<TRecord> records,
+        UpsertRecordOptions? options = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(records);
+
+        foreach (var record in records)
+        {
+            yield return await this.UpsertAsync(record, options, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    #region private
+
+    private async Task CreateIndexAsync(string collectionName, CancellationToken cancellationToken)
+    {
+        var indexCursor = await this._mongoCollection.Indexes.ListAsync(cancellationToken).ConfigureAwait(false);
+        var indexes = indexCursor.ToList(cancellationToken).Select(index => index["name"].ToString()) ?? [];
+        var uniqueIndexes = new HashSet<string?>(indexes);
```

Comment:
Since we just created the collection before calling `CreateIndexAsync`, is it possible for indexes to exist already?  Does indexes for a collection get deleted when the collection is deleted?

Reply:
> Since we just created the collection before calling CreateIndexAsync, is it possible for indexes to exist already?

This is for cases when collection with some indexes already exists in MongoDB, but we want to add new index and avoid creating the same index twice. This is useful when we have a C# type with one vector, and then we modify a type and add another vector property.

This logic exists in current memory store implementation in `main` branch, and I was wondering, if I will remove check for existing indexes, if there will be any errors. Based on my tests, it looks like both `CreateCollection` and `CreateIndex` methods are idempotent, so we can run them multiple times and duplicated collections/indexes won't be created. So, I guess this check just prevents the request to create an index if no index should be created.

> Do indexes for a collection get deleted when the collection is deleted?

Yes.